client/js/render.js

```diff
@@ -19,22 +22,50 @@ module.exports = {
 	renderNetworks,
 };
 
-function buildChannelMessages(channel, messages) {
-	return messages.reduce(function(docFragment, message) {
-		docFragment.append(buildChatMessage({
-			chan: channel,
+function buildChannelMessages(data) {
+	return data.messages.reduce(function(docFragment, message) {
+		appendMessage(docFragment, data.id, data.type, message.type, buildChatMessage({
+			chan: data.id,
 			msg: message
 		}));
 		return docFragment;
 	}, $(document.createDocumentFragment()));
 }
 
+function showAll() {
+	return options.join && options.part && options.quit && options.nick && options.mode;
+}
+
+function appendMessage(container, chan, chanType, messageType, msg) {
+	if (constants.condensedTypes.indexOf(messageType) !== -1 && chanType !== "lobby") {
+		var condensedTypesClasses = "." + constants.condensedTypes.join(", .");
+		var lastChild = container.children("div.msg").last();
+		var lastDate = (new Date(lastChild.attr("data-time"))).toDateString();
+		var msgDate = (new Date(msg.attr("data-time"))).toDateString();
+		if (lastChild && $(lastChild).hasClass("condensed") && !$(msg).hasClass("message") && lastDate === msgDate) {
+			lastChild.append(msg);
+			condense.updateText(lastChild, [messageType]);
+		} else if (lastChild && $(lastChild).is(condensedTypesClasses) && showAll()) {
+			var condensed = buildChatMessage({msg: {type: "condensed", time: msg.attr("data-time")}, chan: chan});
+			condensed.append(lastChild);
+			condensed.append(msg);
+			container.append(condensed);
+			condense.updateText(condensed, [messageType, lastChild.attr("data-type")]);
+		} else {
+			container.append(msg);
+		}
+	} else {
+		container.append(msg);
+	}
+}
+
 function buildChatMessage(data) {
 	const type = data.msg.type;
 	let target = "#chan-" + data.chan;
 	if (type === "error") {
 		target = "#chan-" + chat.find(".active").data("id");
 	}
+	data.msg.previews = data.msg.previews || [];
```

Comment:
This shouldn't be necessary as `previews` is always defined as `[]` on the server.